include/llvm-dialects/Dialect/OpMap.h

```diff
@@ -0,0 +1,356 @@
+    return m_coreOpcodes.contains(op) || m_intrinsics.contains(op);
+  }
+
+  // Check if the map contains an OpDescription created for a given dialect
+  // operation type.
+  template <typename OpT> bool contains() const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return contains(desc);
+  }
+
+  // Check if the map contains an op described by an OpDescription.
+  bool contains(const OpDescription &desc) const {
+    if (desc.isCoreOp() || desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (contains(op))
+          return true;
+      }
+    }
+
+    for (const DialectOpKV &dialectOpKV : m_dialectOps) {
+      if (dialectOpKV == desc)
+        return true;
+    }
+
+    return false;
+  }
+
+  // -------------------------------------------------------------
+  // get returns a reference to the value stored in any of the
+  // internal map containers or constructs a new value.
+  // -------------------------------------------------------------
+  template <typename OpT> ValueT &get() {
+    return get(OpDescription::get<OpT>());
+  }
+
+  // Return a reference to the operation described by a given OpDescription.
+  // This includes any of the possible operation kinds.
+  // Constructs a new entry if the operation was found in neither map.
+  ValueT &get(const OpDescription &desc) {
+    if (desc.isCoreOp()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (m_coreOpcodes.contains(op))
+          return m_coreOpcodes[op];
+      }
+    }
+
+    if (desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (m_intrinsics.contains(op))
+          return m_intrinsics[op];
+      }
+    }
+
+    if (desc.isDialectOp()) {
+      for (DialectOpKV &dialectOpKV : m_dialectOps) {
+        if (dialectOpKV == desc)
+          return dialectOpKV.Value;
+      }
+    }
+
+    // Nothing found. Insert a new entry in the correct map.
+    return insert_or_get(desc, {});
+  }
+
+  // -------------------------------------------------------------
+  // Convenience operator[] definitions.
+  // -------------------------------------------------------------
+  ValueT &operator[](const IntrinsicInst &inst) {
+    return m_intrinsics[inst.getIntrinsicID()];
+  }
+
+  ValueT &operator[](const OpDescription &desc) { return get(desc); }
+
+  // -------------------------------------------------------------
+  // lookup tries to find whether a given function or instruction
+  // can be mapped to any of the entries in the internal map
+  // containers.
+  // It returns either a default-constructed object if the key
+  // was not found or a copy of the contained value.
+  // -------------------------------------------------------------
+
+  template <typename OpT> ValueT lookup(const Function &func) const {
+    return lookup(OpDescription::get<OpT>(), func);
+  }
+
+  ValueT lookup(const OpDescription &desc, const Function &func) const {
```

Comment:
What is this and similar overloads good for? If we already had an `OpDescription`, then we could just lookup based on that, couldn't we? I suspect they should be removed.